alternatives/dotnet/discogs/Parser.cs

```diff
@@ -0,0 +1,102 @@
+using System.Xml;
+using System.Xml.Serialization;
+
+namespace discogs
+{
+    public class Parser<T>
+            where T : IExportToCsv, new()
+    {
+        private readonly int _throttle = 1;
+        private readonly string _typeName;
+        private readonly IExporter<T> _exporter;
+        public Parser(IExporter<T> exporter, int throttle = 1)
+        {
+            _exporter = exporter;
+            _throttle = throttle;
+            _typeName = typeof(T).Name.Split('.')[^1];
+        }
+
+        public event EventHandler<ParseEventArgs> OnSucessfulParse = delegate {};
+
+        public async Task ParseFileAsync(string fileName) {
+            using FileStream fileStream = new FileStream(fileName, FileMode.Open);
+            Stream readingStream = fileStream;
+            if (System.IO.Path.GetExtension(fileName).Equals(".gz", StringComparison.OrdinalIgnoreCase))
+            {
+                readingStream = new GZipStream(fileStream, CompressionMode.Decompress);
+            }
+            await ParseStreamAsync(readingStream);
+        }
+
+        public async Task ParseStreamAsync(Stream stream)
+        {
+            int objectCount = 0;
+            var settings = new XmlReaderSettings
+            {
+                ConformanceLevel = ConformanceLevel.Fragment,
+                Async = true,
+            };
+            using XmlReader reader = XmlReader.Create(stream, settings);
+
+            await reader.MoveToContentAsync();
+            await reader.ReadAsync();
+            while (!reader.EOF)
+            {
+                if (reader.Name == _typeName)
+                {
+                    var objectString = await reader.ReadOuterXmlAsync();
+                    // var objectString = await reader.ReadInnerXmlAsync();
+                    try
+                    {
+                        if (!string.IsNullOrEmpty(objectString))
+                        {
+                            await ExportRecord(objectString);
+                        }
+                        else
+                        {
+                            continue;
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        Console.WriteLine($"Error {ex} parsing node {objectString}");
+                    }
+                    objectCount++;
+                    // await reader.SkipAsync();
+                    if (objectCount % _throttle == 0) OnSucessfulParse(null, new ParseEventArgs { ParseCount = objectCount });
+                    continue;
+                }
+                else
+                {
+                    await reader.ReadAsync();
+                }
+            }
+            await _exporter.CompleteExportAsync(objectCount);
+        }
+
+        protected static T Deserialize(string content)
+        {
+            using var reader = new StringReader(content);
+            XmlSerializer _labelXmlSerializer = new XmlSerializer(typeof(T));
```

Comment:
Moving this serializer instance outside (into a static field) will make this code a lot faster.

Based on its name, it seems that this used to be the case?!?

Reply:
Moved to static field.